### PR TITLE
refactor: release_dag

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- To accommodate the decomposition base, the calculation of the node's deadline at the time of DAG release is branched.
- - Fixed deadline_factor at 10^5 for comparison.

- node_integer_absolute_deadline = integer_scaled_deadline + integer_scaled_offset
- absolute_deadline = node_integer_absolute_deadline * release_count or dag_period * release_count

## Related links

None.

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: lib/src/dag_set_scheduler.rs [hunk 1]
```diff
@@ -109,10 +109,25 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
...
-                    "absolute_deadline",
-                    dag.get_head_period().unwrap() * managers[dag_id].get_release_count(),
-                );
+                // If Node does not have individual deadlines, use DAG deadline.
+                if dag[NodeIndex::new(0)]
+                    .params
+                    .contains_key("int_scaled_node_relative_deadline")
+                {
+                    for node_i in dag.node_indices() {
+                        let node_deadline =
```
Comment (atsushi421) [inline lib/src/dag_set_scheduler.rs:118]:
```suggestion
                        let node_relative_deadline =
```

File: lib/src/dag_set_scheduler.rs [hunk 2]
```diff
@@ -109,10 +109,25 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
...
+                            dag[node_i].get_params_value("int_scaled_node_relative_deadline");
+                        dag[node_i].params.insert(
+                            "int_scaled_node_absolute_deadline".to_string(),
+                            node_deadline * managers[dag_id].get_release_count(),
+                        );
+                    }
+                } else {
+                    dag.set_dag_param(
+                        "absolute_deadline",
```
Comment (atsushi421) [inline lib/src/dag_set_scheduler.rs:127]:
```suggestion
                        "node_absolute_deadline",
```

File: lib/src/dag_set_scheduler.rs [hunk 3]
```diff
@@ -109,10 +109,25 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
...
+                        dag.get_head_period().unwrap() * managers[dag_id].get_release_count(),
```
Comment (atsushi421) [inline lib/src/dag_set_scheduler.rs:128]:
```suggestion
                        dag.get_end_to_end_deadline().unwrap() * managers[dag_id].get_release_count(),
```

File: lib/src/global_edf_scheduler.rs [hunk 4]
```diff
@@ -12,7 +12,20 @@ use crate::{
...
+        if self
+            .node_data
+            .params
+            .contains_key("int_scaled_node_absolute_deadline")
+            && other
+                .node_data
+                .params
+                .contains_key("int_scaled_node_absolute_deadline")
+        {
+            comparison_metric = "int_scaled_node_absolute_deadline";
```
Comment (atsushi421) [inline lib/src/global_edf_scheduler.rs:26]:
```suggestion
            comparison_metric = "int_scaled_node_absolute_deadline";  // decompsition-based algorithm
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 5]
```diff
@@ -1,61 +1,54 @@
...
-
 #[allow(dead_code)]
 pub fn decompose(dag: &mut Graph<NodeData, i32>) {
     let mut segments = create_segments(dag);
     calculate_segments_deadline(dag, &mut segments);
 
-    // Add deadlines to nodes.
-    let mut nodes_deadline = vec![0.0; dag.node_count()];
+    // Calculate integer scaled deadlines and offsets.
+    let deadline_factor = 100000.0;
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs:13]: Changed.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 6]
```diff
@@ -1,61 +1,54 @@
...
-            .to_string()
-            .chars()
-            .skip(integer_part_str.len() + 1)
-            .collect::<String>();
-        let deadline_factor = 10u64.pow(fractional_part_str.len().try_into().unwrap()) as i32;
-        let node_i = NodeIndex::new(i);
-        dag.add_param(node_i, "deadline_factor", deadline_factor);
+    let int_scaled_offset = calc_int_scaled_offsets(dag, &int_scaled_deadline);
+
+    // Set integer scaled absolute deadline.
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs]: Changed.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs [hunk 7]
```diff
@@ -91,17 +84,11 @@ mod tests {
         let mut dag = create_sample_dag(120);
         decompose(&mut dag);
 
-        let expect_deadline = [322857, 1033721, 7318570, 5316279, 4358571];
-        let expect_offset = [0, 322857, 322857, 1356578, 7641427];
+        let expect_absolute_deadline = [322857, 1356578, 7641428, 6672857, 11999999];
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs]: Changed.

File: lib/src/dag_set_scheduler.rs (hunk 1 above)
Comment (kobayu858) [reply lib/src/dag_set_scheduler.rs]: Changed.

File: lib/src/dag_set_scheduler.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/dag_set_scheduler.rs]: Changed.

File: lib/src/dag_set_scheduler.rs (hunk 3 above)
Comment (kobayu858) [reply lib/src/dag_set_scheduler.rs]: Changed.

File: lib/src/global_edf_scheduler.rs (hunk 4 above)
Comment (kobayu858) [reply lib/src/global_edf_scheduler.rs]: Changed.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**